### PR TITLE
[frontend/backend] Add triggers management in Group overview settings (#3520)

### LINKED ISSUE #3520: In Settings, be able to display and manage triggers & digests for groups / organizations
Create/display/update/delete  triggers & digests  for a group.

Create/display/update/delete  triggers & digests  for a Organizations
--> Need an Overview for Organizations

See Notion for details

### PR DESCRIPTION
**Pull request by marieflorescontact**
… (#3520)

### Proposed changes

* Refactor `triggerLineTitles` => iterate on dataColumns 
* Add Trigger and Digest list in Settings > Group
* Add `currentUserAccessrigh`t in `TriggerLine` => display or not the update/delete popover depending on user access right on Trigger
* `Schema` changes => Add `group_ids` filter , remove `user_ids` field and add `currentUserAccessright`
* User has `view` access_right with a group Trigger  but `admin` access_right if user trigger 
* Use `authorized_members` instead of `group_ids` and `user_ids` in `notificationManager`
* Migration => remove unused elastic field `user_id`s, `group_ids`, `recipients`
* Tests ❤️

### Related issues

* [issue/3520](https://github.com/OpenCTI-Platform/opencti/issues/3520)

### Checklist

- [x] I consider the submitted work as finished
- [x] I tested the code for its functionality
- [x] I wrote test cases for the relevant uses case
- [x] I added/update the relevant documentation (either on github or on notion) WIP
- [x] Where necessary I refactored code to improve the overall quality

### Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: opencti-platform/opencti-graphql/src/modules/notification/notification-domain.ts [hunk 1]
```diff
@@ -48,22 +49,38 @@ const extractUniqRecipient = async (
...
       throw ForbiddenAccess();
     }
     if (triggerInput.recipients.length > 1) {
       throw UnsupportedError(`Cannot create ${type} trigger for more than one recipient`);
     }
+    // TODO refacto
     const [recipientId] = triggerInput.recipients;
     const recipientUser = await internalLoadById<BasicStoreEntity>(context, user, recipientId, { type: ENTITY_TYPE_USER });
-    if (!recipientUser) {
+    const recipientGroup = await internalLoadById<BasicStoreEntity>(context, user, recipientId, { type: ENTITY_TYPE_GROUP });
```
Comment (SouadHadjiat) [inline opencti-platform/opencti-graphql/src/modules/notification/notification-domain.ts]: Suggestion : load the entity by id without the `type` filter, and check if the returned entity `entity_type` is User or Group. The recipient will already contain the entity_type so no need to create another object @marieflorescontact

File: opencti-platform/opencti-front/src/private/components/profile/TriggerLineTitles.tsx [hunk 2]
```diff
@@ -33,36 +33,15 @@ const TriggerLineTitles: FunctionComponent<TriggerLineTitlesProps> = ({
...
-            >
-              {t(dataColumns.event_types.label)}
-            </div>
-            <div
-              className={classes.headerItem}
-              style={{ width: dataColumns.filters.width }}
-            >
-              {t(dataColumns.filters.label)}
-            </div>
+            {Object.values(dataColumns).map((value) => (
```
Comment (Archidoit) [inline opencti-platform/opencti-front/src/private/components/profile/TriggerLineTitles.tsx:36]: Can you put this code in a generic component, that would be used instead of TriggerLineTitles, UserLineTitles and TaskLineTitles?

File: opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx [hunk 3]
```diff
@@ -173,6 +189,53 @@ const Group = ({ groupData }: { groupData: Group_group$key }) => {
     (group.default_marking ?? []).find((d) => d.entity_type === 'GLOBAL')
       ?.values ?? [],
   );
+
+  // Triggers and Digest
+  const paginationOptions = {
+    count: 25,
+    filters: [{ key: ['group_ids' as TriggerFilter], values: [group.id] }],
```
Comment (Archidoit) [inline opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx]: why do you need to precise the type TriggerFilter here?

File: opencti-platform/opencti-graphql/tests/02-integration/02-resolvers/trigger-test.js [hunk 4]
```diff
@@ -118,11 +119,54 @@ describe('Trigger resolver standard behavior', () => {
...
-    const queryResult = await queryAsAdmin({ query: LIST_QUERY, variables: { first: 10 } });
-    expect(queryResult.data.triggers.edges.length).toEqual(2);
+  // Group trigger start
+  it('security user should create group trigger', async () => {
+    // Create the trigger
+    const GROUP_TRIGGER_TO_CREATE = {
+      input: {
+        name: 'group trigger',
+        description: '',
+        event_types: 'create',
```
Comment (Archidoit) [inline opencti-platform/opencti-graphql/tests/02-integration/02-resolvers/trigger-test.js]: EVENT_TYPE_CREATE instead of 'create' to use the constant

**Comment by Archidoit** [comment]
maybe add more space before the 'live trigger' icon
image

File: opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx [hunk 5]
```diff
@@ -344,6 +407,64 @@ const Group = ({ groupData }: { groupData: Group_group$key }) => {
               </Grid>
             </Paper>
           </Grid>
+          <Grid item={true} xs={12} style={{ marginTop: 30 }}>
+            <Typography
+              variant="h4"
+              gutterBottom={true}
+              style={{ float: 'left' }}
+            >
+              {t('Triggers and Digests')}
```
Comment (Archidoit) [inline opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx]: no upper case for the first letter of 'digest' in the translations

**Comment by Archidoit** [comment]
missing translations: 'Add a live trigger' and 'Add a regular digest' (tooltips)
image

File: opencti-platform/opencti-front/src/private/components/profile/triggers/TriggerLine.tsx [hunk 6]
```diff
@@ -247,11 +248,19 @@ export const TriggerLineComponent: FunctionComponent<TriggerLineProps> = ({
                 )}
               </div>
             )}
+            <div
+              className={classes.bodyItem}
+              style={{ width: dataColumns.ownership.width }}
+            >
+              {data.currentUserAccessRight === 'admin' ? t('User') : t('Group')}
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/profile/triggers/TriggerLine.tsx]: I don't understand this condition
If I create a trigger in Group, the owner is the user
So, if an other user is in this Group, the owner should be this main user no ?

File: opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx [hunk 7]
```diff
@@ -344,6 +407,64 @@ const Group = ({ groupData }: { groupData: Group_group$key }) => {
               </Grid>
             </Paper>
           </Grid>
+          <Grid item={true} xs={12} style={{ marginTop: 30 }}>
+            <Typography
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx]: I think, it will be good to have a dedicated component for triggers and digests to be mutualize with code in User.tsx component

File: opencti-platform/opencti-graphql/src/migrations/1687529127415-triggers-remove-unused-field.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+      }
+    }
+  };
+  const message = '[MIGRATION] Triggers remove unused fields: recipients, user_ids, group_ids';
+  return elUpdateByQueryForMigration(message, READ_INDEX_INTERNAL_OBJECTS, updateQuery).catch((err) => {
+    throw DatabaseError('Error updating elastic', { error: err });
+  });
+};
+export const up = async (next) => {
+  await updateTriggers();
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-graphql/src/migrations/1687529127415-triggers-remove-unused-field.js:49]: Why do we have a method instead of code ?

File: opencti-platform/opencti-graphql/src/modules/notification/notification-domain.ts [hunk 9]
```diff
@@ -48,22 +50,19 @@ const extractUniqRecipient = async (
...
-  if (triggerInput.recipients && triggerInput.recipients.length > 0) {
+): Promise<BasicStoreEntity> => {
+  const { recipients } = triggerInput;
+  let recipient = user.id;
+  if (recipients?.at(0)) {
     if (!isUserHasCapability(user, SETTINGS_SET_ACCESSES)) {
       throw ForbiddenAccess();
     }
-    if (triggerInput.recipients.length > 1) {
+    if (recipients?.at(1)) {
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-graphql/src/modules/notification/notification-domain.ts]: Why this changes ?
isn't that more readable like this -> recipients?.length > 1 ?

File: opencti-platform/opencti-graphql/src/modules/notification/notification-domain.ts [hunk 10]
```diff
@@ -48,22 +50,19 @@ const extractUniqRecipient = async (
...
       throw UnsupportedError(`Cannot create ${type} trigger for more than one recipient`);
     }
-    const [recipientId] = triggerInput.recipients;
-    const recipientUser = await internalLoadById<BasicStoreEntity>(context, user, recipientId, { type: ENTITY_TYPE_USER });
-    if (!recipientUser) {
-      throw UnsupportedError('Recipient user not found');
-    }
-    return recipientUser;
+    recipient = recipients.at(0) as string;
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-graphql/src/modules/notification/notification-domain.ts]: You can avoid the as string

    recipient = recipients[0];

File: opencti-platform/opencti-graphql/src/modules/notification/notification-domain.ts [hunk 11]
```diff
@@ -127,15 +134,24 @@ export const triggerDelete = async (context: AuthContext, user: AuthUser, trigge
 };
 export const triggersFind = (context: AuthContext, user: AuthUser, opts: QueryTriggersArgs) => {
   let queryArgs = { ...opts };
-  if (opts.filters?.some((f) => f.key.includes(TriggerFilter.UserIds))) {
+  // key is a string[] because of the resolver, we have updated the keys
+  const userIdFilter = opts.filters?.find((f) => (f.key as string[]).includes('authorized_members.id'));
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-graphql/src/modules/notification/notification-domain.ts:137]: I don't understand this line
We don't have authorized_members as filter key in GraphQL

File: opencti-platform/opencti-graphql/src/modules/notification/notification-domain.ts [hunk 12]
```diff
@@ -127,15 +134,24 @@ export const triggerDelete = async (context: AuthContext, user: AuthUser, trigge
...
-    queryArgs = { ...queryArgs, adminBypassUserAccess: true };
+    queryArgs = {
+      ...queryArgs,
+      adminBypassUserAccess: true
+    };
   }
   return listEntitiesPaginated<BasicStoreEntityTrigger>(context, user, [ENTITY_TYPE_TRIGGER], queryArgs);
 };
 
+export const getCurrentUserAccessRight = async (context: AuthContext, user: AuthUser, trigger: BasicStoreEntityTrigger) => {
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-graphql/src/modules/notification/notification-domain.ts]: This method seems to be not used

File: opencti-platform/opencti-graphql/src/modules/notification/notification-resolver.ts [hunk 13]
```diff
@@ -32,6 +33,11 @@ const notificationResolvers: Resolvers = {
   },
   Trigger: {
     triggers: (trigger, _, context) => triggersGet(context, context.user, trigger.trigger_ids),
+    currentUserAccessRight: (trigger, _, context) => getUserAccessRight(context.user, trigger),
+  },
+  TriggerFilter: {
+    user_ids: 'authorized_members.id',
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-graphql/src/modules/notification/notification-resolver.ts:39]: Why we don't have directly authorized_members.id in filter key ?

**Comment by RomuDeuxfois** [comment]
Don't forget to feed the [documentation](https://docs.opencti.io/5.8.X/administration/introduction/) :)

File: opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx (hunk 3 above)
Comment (marieflorescontact) [reply opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx]: To avoid typescript error:  `Type 'string' is not assignable to type 'TriggerFilter'.`

File: opencti-platform/opencti-front/src/private/components/profile/TriggerLineTitles.tsx (hunk 2 above)
Comment (marieflorescontact) [reply opencti-platform/opencti-front/src/private/components/profile/TriggerLineTitles.tsx:36]: after discussion with @Kedae , it is not necessary for the moment

File: opencti-platform/opencti-graphql/tests/02-integration/02-resolvers/trigger-test.js (hunk 4 above)
Comment (marieflorescontact) [reply opencti-platform/opencti-graphql/tests/02-integration/02-resolvers/trigger-test.js]: fixed

File: opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx (hunk 5 above)
Comment (marieflorescontact) [reply opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx]: fixed

File: opencti-platform/opencti-front/src/private/components/profile/triggers/TriggerLine.tsx (hunk 6 above)
Comment (marieflorescontact) [reply opencti-platform/opencti-front/src/private/components/profile/triggers/TriggerLine.tsx]: as discussed with @Kedae i'll change the wording for `Private` and `Shared`

Comment (Kedae) [reply opencti-platform/opencti-front/src/private/components/profile/triggers/TriggerLine.tsx]: Private and Shared :)

File: opencti-platform/opencti-graphql/src/modules/notification/notification-resolver.ts (hunk 13 above)
Comment (marieflorescontact) [reply opencti-platform/opencti-graphql/src/modules/notification/notification-resolver.ts:39]: because `authorized_members` is not functionally comprehensible

File: opencti-platform/opencti-front/src/private/components/profile/triggers/TriggerLine.tsx (hunk 6 above)
Comment (marieflorescontact) [reply opencti-platform/opencti-front/src/private/components/profile/triggers/TriggerLine.tsx]: we won't have anymore column but a grayed out burger button with a tooltip "This trigger/digest has been shared with you an you are not able to modify or delete it."

File: opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx (hunk 7 above)
Comment (marieflorescontact) [reply opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx]: I fully agree. i'm working on it.

File: opencti-platform/opencti-graphql/src/migrations/1687529127415-triggers-remove-unused-field.js (hunk 8 above)
Comment (marieflorescontact) [reply opencti-platform/opencti-graphql/src/migrations/1687529127415-triggers-remove-unused-field.js:49]: Not sure to understand your comment: the method called is defined above the `up`

File: opencti-platform/opencti-graphql/src/modules/notification/notification-domain.ts (hunk 10 above)
Comment (marieflorescontact) [reply opencti-platform/opencti-graphql/src/modules/notification/notification-domain.ts]: This is to avoid Typescript error :   `Type 'undefined' is not assignable to type 'string'` ( `recipients.at(0)` can return undefined)

File: opencti-platform/opencti-graphql/src/modules/notification/notification-domain.ts (hunk 11 above)
Comment (marieflorescontact) [reply opencti-platform/opencti-graphql/src/modules/notification/notification-domain.ts:137]: `authorized_members` is the filter in elastic, but it is not a is not functionally comprehensible. So in the api we use `group_ids` and `user_ids` (resolver, line 38) :
```
  TriggerFilter: {
    user_ids: 'authorized_members.id',
    group_ids: 'authorized_members.id',
  },
```

File: opencti-platform/opencti-graphql/src/modules/notification/notification-domain.ts (hunk 12 above)
Comment (marieflorescontact) [reply opencti-platform/opencti-graphql/src/modules/notification/notification-domain.ts]: fixed

File: opencti-platform/opencti-graphql/src/modules/notification/notification-domain.ts (hunk 10 above)
Comment (Kedae) [reply opencti-platform/opencti-graphql/src/modules/notification/notification-domain.ts]: I think you can try [recipient] = recipients;

Comment (marieflorescontact) [reply opencti-platform/opencti-graphql/src/modules/notification/notification-domain.ts]: fixed( and no Typescript error)

File: opencti-platform/opencti-graphql/src/modules/notification/notification-domain.ts (hunk 9 above)
Comment (marieflorescontact) [reply opencti-platform/opencti-graphql/src/modules/notification/notification-domain.ts]: fixed

File: opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx (hunk 7 above)
Comment (marieflorescontact) [reply opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx]: done

**Comment by marieflorescontact** [comment]
**Post demo comments:**:

- Need to move Triggers list in settings before `Members` ==>done
- Need to fix horizontal scrollers in Triggers list ===> done

File: opencti-platform/opencti-front/src/private/components/profile/triggers/TriggerLine.tsx [hunk 14]
```diff
@@ -132,6 +134,11 @@ export const TriggerLineComponent: FunctionComponent<TriggerLineProps> = ({
   const time = currentTime.length > 1
     ? new Date(`2000-01-01T${currentTime[1]}`)
     : new Date(`2000-01-01T${currentTime[0]}`);
+
+  const hasAccess = () => {
```
Comment (Kedae) [inline opencti-platform/opencti-front/src/private/components/profile/triggers/TriggerLine.tsx]: No need for a function here

Comment (marieflorescontact) [reply opencti-platform/opencti-front/src/private/components/profile/triggers/TriggerLine.tsx]: fixed

**Review by richard-julien — APPROVED** [review]
